[dnf5] Add rpm query class

j-mracek opened the pull request.

j-mracek committed:
[dnf5] Add rpm query class with first filters

jrohel commented on include/libdnf/rpm/query.hpp:
```diff
@@ -0,0 +1,75 @@
+
+private:
+    class Impl;
+    std::unique_ptr<Impl> pImpl;
```
May be `p_impl` - we agreed on snake_case variables.

jrohel commented on libdnf/rpm/sack_impl.hpp:
```diff
@@ -20,13 +20,39 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+#include <vector>
+
+constexpr const char * SOLVABLE_NAME_ADVISORY_PREFIX = "patch:";
+constexpr size_t SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH = 6;
```
We need `SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH` constant?
Anyway compiler (during compile time, no cost during runtime) can compute the length. There is no need to hardcode the value by hand:
`constexpr size_t SOLVABLE_NAME_ADVISORY_PREFIX_LENGTH = std::char_traits<char>::length(SOLVABLE_NAME_ADVISORY_PREFIX);`

j-mracek committed:
[dnf5] Use PackageId in SolvMap

PackageId provides type specific structure instead of int/Id.

j-mracek committed:
[dnf5] Rename files according to class name

j-mracek committed:
[dnf5] Add ifilters for SolvQuery

j-mracek committed:
[dnf5] Rename class Query to SolvQuery

Query class was alreadu use as a template for other classes.

j-mracek committed:
[dnf5] Add additional ifilters for SolvQuery

j-mracek committed:
[dnf5] Replace if by switch plus clang changes

j-mracek committed:
[dnf5] Replace part of the code by template

j-mracek committed:
[dnf5] Add ifilter for SolvQuery + format (clang) changes

Additionally there are minor changes created by clang.

j-mracek committed:
[dnf5] Changes required by rebase

j-mracek committed:
[dnf5] Add method size for SolvQuery

j-mracek committed:
Add get_packageset() method for SolvQuery

jrohel commented:
@j-mracek I'm trying to use the your code. 
Implementation of `rpm::SolvQuery destructors` are missing. For my test I added this into solv_query.cpp:
```
SolvQuery::Impl::~Impl() = default;
SolvQuery::~SolvQuery() = default;
```
I know your work is in progress. So, it's only a tip.

j-mracek committed:
[dnf5] Add default destructor for SolvQuery

j-mracek committed:
[dnf5] Add unittest for SolvQuery.ifilter_name

j-mracek committed:
[dnf5] Add get_full_nevra for Package class

The original function returned epoch only whet it was different form 0.
The new function returns always EPOCH.

jrohel commented on include/libdnf/rpm/solv_query.hpp:
```diff
@@ -0,0 +1,174 @@
+
+    struct NotSupportedCmpType : public RuntimeError {
+        using RuntimeError::RuntimeError;
+        const char * get_domain_name() const noexcept override { return "libdnf::rpm::Query"; }
```
`"libdnf::rpm::SolvQuery"`

j-mracek committed:
[dnf5] Do not create new id in pool in ifilter_evr()

Creation of permanent Id in pool could significantly enlarge the pool.

j-mracek committed:
Do not create new pool id for evr during NEVRA parsing

It prevents enlargement of pool and creation of permanent Ids.

j-mracek committed:
Rename SolvQuery.ifilter_nevra_strict

In dnf4 there are two filters - nevra and nevra_strict but in the new
branch there will be only one nevra filter. I propose the shorter name
as a better option.

j-mracek committed:
Strip epoch by a way like libsolv

It will unify evr strings format before version comparison.

j-mracek committed:
Add unittests for SolvQuery

It also fixes format of comments (/// replaced by //)

j-mracek committed:
Rename id_map query_result in SolvQuery

j-mracek committed:
Add code that disappeared during rebase

jrohel commented:
Commit messages:
Add ifilter for SolvQuery -> ifilter for SolvQuery + format (clang) changes
Rename SolvQuery.ifilter_nevra_strict :  sorter -> shorter

jrohel commented on test/libdnf/rpm/test_solv_query.cpp:
```diff
@@ -164,4 +170,15 @@ void RpmSolvQueryTest::test_ifilter_name() {
     CPPUNIT_ASSERT_THROW(query8.ifilter_name(libdnf::sack::QueryCmp::GT, names_icontains);,
                          libdnf::rpm::SolvQuery::NotSupportedCmpType);
+
+    /// Test QueryCmp::EQ with two elements
```
On more places:
Is it a string for generating documentation?
Comment - `//`
Documentation string - `///`

j-mracek replied:
Only comment, I will change it.

j-mracek replied:
Fixed

jrohel commented on include/libdnf/rpm/solv_query.hpp:
```diff
@@ -0,0 +1,174 @@
+#include <string>
+#include <vector>
+
+//#include <string>
```
Forgotten line?

j-mracek replied:
Fixed

jrohel commented on libdnf/rpm/package_set_impl.hpp:
```diff
@@ -62,8 +62,8 @@ inline PackageSet::Impl::Impl(SolvSack * sack)
-inline PackageSet::Impl::Impl(SolvSack * sack, Map * map)
-    : libdnf::rpm::solv::SolvMap::SolvMap(map)
+inline PackageSet::Impl::Impl(SolvSack * sack, libdnf::rpm::solv::SolvMap & solv_map)
+    : libdnf::rpm::solv::SolvMap::SolvMap(solv_map)
```
We are in libdnf::namespace:
`libdnf::rpm::solv::SolvMap::SolvMap(solv_map)` -> `solv::SolvMap::SolvMap(solv_map)`

j-mracek replied:
> Commit messages:
> Add ifilter for SolvQuery -> ifilter for SolvQuery + format (clang) changes
> Rename SolvQuery.ifilter_nevra_strict : sorter -> shorter

Fixed

jrohel commented on libdnf/rpm/solv/package_private.hpp:
```diff
@@ -188,6 +192,54 @@ inline const char * get_nevra(Pool * pool, libdnf::rpm::PackageId package_id) no
 }
 
+/// @return const char* !! Return temporal value !!
+inline const char * get_full_nevra(Pool * pool, libdnf::rpm::PackageId package_id) noexcept {
```
Good candidate to move to .cpp (to be not inline function, big function).

j-mracek committed:
Remove unneeded namespace prefix

In namespace libdnf::rpm it is not needed to use libdnf::rpm::solv::,
but only solv::.

j-mracek committed:
Do not use inline for a large function and removal of noexcept

The function could also allocate a temporal memory in pool.

jrohel commented:
Thanks.
